analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs

```diff
@@ -18,161 +18,167 @@
             {
-                if (!invocation.ArgumentList.Arguments.Any())
+                if (methodSymbol.Name == "OfType" && CanHaveNullValue(castType))
                 {
-                    return null;
+                    // OfType() filters 'null' values from enumerables
+                    return;
                 }
-                collection = invocation.ArgumentList.Arguments.First().Expression;
-            }
-            else
-            {
-                if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+
+                var elementType = GetElementType(invocation, methodSymbol, context.SemanticModel);
+                // Generic types {T} and {T?} are equal and there is no way to access NullableAnnotation field right now
+                // See https://github.com/SonarSource/sonar-dotnet/issues/3273
+                if (elementType != null && elementType.Equals(castType) && string.Equals(elementType.ToString(), castType.ToString(), StringComparison.Ordinal))
                 {
-                    return null;
+                    var methodCalledAsStatic = methodSymbol.MethodKind == MethodKind.Ordinary;
+                    ReportIssue(context, invocation, GetReportLocation(invocation, methodCalledAsStatic), returnType);
                 }
-                collection = memberAccess.Expression;
             }
+        }
+    }
 
-            var typeInfo = semanticModel.GetTypeInfo(collection);
-            if (typeInfo.Type is INamedTypeSymbol collectionType &&
-                collectionType.TypeArguments.Length == 1)
+    private static void ReportIssue(SonarSyntaxNodeReportingContext context,
+                                    ExpressionSyntax expression,
+                                    Location location,
+                                    ITypeSymbol castType) =>
+        context.ReportIssue(Diagnostic.Create(Rule, location, castType.ToMinimalDisplayString(context.SemanticModel, expression.SpanStart)));
+
+    /// If the invocation one of the <see cref="CastIEnumerableMethods"/> extensions, returns the method symbol.
+    private static IMethodSymbol GetEnumerableExtensionSymbol(InvocationExpressionSyntax invocation, SemanticModel semanticModel) =>
+        invocation.GetMethodCallIdentifier() is { } methodName
+        && CastIEnumerableMethods.Contains(methodName.ValueText)
+        && semanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol
+        && methodSymbol.IsExtensionOn(KnownType.System_Collections_IEnumerable)
+            ? methodSymbol
+            : null;
+
+    private static ITypeSymbol GetGenericTypeArgument(ITypeSymbol type) =>
+        type is INamedTypeSymbol returnType && returnType.Is(KnownType.System_Collections_Generic_IEnumerable_T)
+            ? returnType.TypeArguments.Single()
+            : null;
+
+    private static bool CanHaveNullValue(ITypeSymbol type) => type.IsReferenceType || type.Name == "Nullable";
```

Comment:
Formatting is of here
```suggestion
    private static bool CanHaveNullValue(ITypeSymbol type) =>
        type.IsReferenceType || type.Name == "Nullable";
```